### PR TITLE
[Merged by Bors] - feat: the universal property of a disjoint union of topological spaces

### PR DESCRIPTION
**Pull request by dagurtomas**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 1]
```diff
@@ -325,12 +325,35 @@ def prodSwap : C(α × β, β × α) := .prodMk .snd .fst
...
+-/
+def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
+  toFun ig := f ig.fst ig.snd
+
+variable (A X) in
+/--
+Giving a continuous map out of a disjoint union is the same as giving a continuous map out of
+each term
+-/
+def sigmaEquiv : (∀ i, C(X i, A)) ≃ C((Σ i, X i), A) where
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:349]: Can you add a reference to `Equiv.piCurry` in this docstring?

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 2]
```diff
@@ -351,6 +374,17 @@ theorem pi_eval (f : ∀ i, C(A, X i)) (a : A) : (pi f) a = fun i : I => (f i) a
 def eval (i : I) : C(∀ j, X j, X i) where
   toFun := Function.eval i
 
+variable (A X) in
+/--
+Giving a continuous map out of a disjoint union is the same as giving a continuous map out of
+each term
+-/
+def piEquiv : (∀ i, C(A, X i)) ≃ C(A, ∀ i, X i) where
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:382]:
```suggestion
@[simps]
def piEquiv : (∀ i, C(A, X i)) ≃ C(A, ∀ i, X i) where
```

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 3]
```diff
@@ -351,6 +374,17 @@ theorem pi_eval (f : ∀ i, C(A, X i)) (a : A) : (pi f) a = fun i : I => (f i) a
 def eval (i : I) : C(∀ j, X j, X i) where
   toFun := Function.eval i
 
+variable (A X) in
+/--
+Giving a continuous map out of a disjoint union is the same as giving a continuous map out of
+each term
+-/
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:381]: This possibly deserves a reference to `Equiv.curry`; I'll let another reviewer decide.

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 4]
```diff
@@ -325,12 +325,35 @@ def prodSwap : C(α × β, β × α) := .prodMk .snd .fst
...
-    C(Y i, Σ i, Y i) where
+def sigmaMk (i : I) : C(X i, Σ i, X i) where
   toFun := Sigma.mk i
 
+/--
+To give a continuous map out of a disjoint union, it suffices to give a continuous map out of
+each term
+-/
+def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
+  toFun ig := f ig.fst ig.snd
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:342]: I think this is
```suggestion
/--
To give a continuous map out of a disjoint union, it suffices to give a continuous map out of
each term
-/
def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
  toFun := Sigma.curry fun i x => f i x
```
which is probably worth mentioning in the docstring, if you don't want this implementation.

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 5]
```diff
@@ -325,12 +325,37 @@ def prodSwap : C(α × β, β × α) := .prodMk .snd .fst
...
-    C(Y i, Σ i, Y i) where
+def sigmaMk (i : I) : C(X i, Σ i, X i) where
   toFun := Sigma.mk i
 
+/--
+To give a continuous map out of a disjoint union, it suffices to give a continuous map out of
+each term. This is `Sigma.uncurry fun i x => f i x`, but `aesop` doesn't manage to prove
+continuity with that implementation.
+-/
+def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:342]:
```suggestion
def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
```
```suggestion
@[simps]
def sigma (f : ∀ i, C(X i, A)) : C((Σ i, X i), A) where
```

File: Mathlib/Topology/ContinuousFunction/Basic.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Topology/ContinuousFunction/Basic.lean:384]: Nevermind, that's not relevant here

File: Mathlib/Topology/ContinuousFunction/Basic.lean [hunk 6]
```diff
@@ -325,12 +325,38 @@ def prodSwap : C(α × β, β × α) := .prodMk .snd .fst
...
 @[simps apply]
-def sigmaMk {ι : Type*} {Y : ι → Type*} [∀ i, TopologicalSpace (Y i)] (i : ι) :
-    C(Y i, Σ i, Y i) where
+def sigmaMk (i : I) : C(X i, Σ i, X i) where
   toFun := Sigma.mk i
 
+/--
+To give a continuous map out of a disjoint union, it suffices to give a continuous map out of
+each term. This is `Sigma.uncurry fun i x => f i x`, but `aesop` doesn't manage to prove
+continuity with that implementation.
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Basic.lean:340]:
```suggestion
each term. This is `Sigma.uncurry fun i x => f i x`, but `aesop` doesn't manage to prove
continuity with that implementation.
```
I think it's enough to say

```suggestion
each term. This is `Sigma.uncurry` for continuous maps.
```

**Review by eric-wieser — APPROVED** [review]
maintainer merge

Thanks!

**Review by jcommelin — APPROVED** [review]
bors d+

**Pull request closed without merging**